src/layer.rs

```diff
@@ -16,34 +16,197 @@ use crate::{Error, Glyph, Plist};
+    pub fn len(&self) -> usize {
+        self.layers.len()
+    }
+
+    /// Get a reference to a layer, by name.
+    pub fn get(&self, name: &str) -> Option<&Layer> {
+        self.layers.iter().find(|l| &*l.name == name)
+    }
+
+    /// Get a mutable reference to a layer, by name.
+    pub fn get_mut(&mut self, name: &str) -> Option<&mut Layer> {
+        self.layers.iter_mut().find(|l| &*l.name == name)
+    }
+
+    /// A reference to the default layer.
+    pub fn default_layer(&self) -> &Layer {
+        &self.layers[0]
+    }
+
+    /// A mutable reference to the default layer.
+    pub fn default_layer_mut(&mut self) -> &mut Layer {
+        &mut self.layers[0]
+    }
+
+    /// Iterate over all layers.
+    pub fn iter(&self) -> impl Iterator<Item = &Layer> {
+        self.layers.iter()
+    }
+
+    /// Iterate over the names of all layers.
+    pub fn names(&self) -> impl Iterator<Item = &LayerName> {
+        self.layers.iter().map(|l| &l.name)
+    }
+
+    /// Create a new layer with the given name.
+    pub fn new_layer(&mut self, name: &str) -> Result<(), Error> {
+        if self.layers.iter().any(|l| &*l.name == name) {
+            Err(Error::DuplicateLayer(name.into()))
+        } else {
+            let layer = Layer::new(name.into(), None);
+            self.layers.push(layer);
+            Ok(())
+        }
+    }
+
+    /// Remove a layer.
+    ///
+    /// The default layer cannot be removed.
+    pub fn remove(&mut self, name: &str) -> Option<Layer> {
+        self.layers.iter().position(|l| &*l.name == name).map(|idx| self.layers.remove(idx))
+    }
+
+    /// Rename a layer.
+    ///
+    /// If `overwrite` is true, and a layer with the new name exists, it will
+    /// be replaced.
+    ///
+    /// Returns an error if `overwrite` is false but a layer with the new
+    /// name exists, or if no layer with the old name exists.
+    pub fn rename_layer(&mut self, old: &str, new: &str, overwrite: bool) -> Result<(), Error> {
+        if !overwrite && self.get(new).is_some() {
+            Err(Error::DuplicateLayer(new.into()))
+        } else if self.get(old).is_none() {
+            Err(Error::MissingLayer(old.into()))
+        } else {
+            if overwrite {
+                self.layers.retain(|l| &*l.name != new)
+            }
+            self.get_mut(old).unwrap().name = new.into();
+            Ok(())
+        }
+    }
+}
+
+impl Default for LayerSet {
+    fn default() -> Self {
+        let layer = Layer::new(DEFAULT_LAYER_NAME.into(), None);
+        let layers = vec![layer];
+        LayerSet { layers }
+    }
+}
+
 /// A [layer], corresponding to a 'glyphs' directory. Conceptually, a layer
 /// is just a collection of glyphs.
 ///
 /// [layer]: http://unifiedfontobject.org/versions/ufo3/glyphs/
-#[derive(Debug, Clone, Default, PartialEq)]
+#[derive(Debug, Clone, PartialEq)]
 pub struct Layer {
     pub(crate) glyphs: BTreeMap<GlyphName, Arc<Glyph>>,
+    pub name: LayerName,
```

Comment:
Doesn't this also let the user rename a layer outside `rename_layer`'s control?

Reply:
Yea, this is a funny case. Layer names and paths are annoying; there's a real sense in which they aren't part of the layer itself, which was the motivation for the original design.

In any case, I think you're right, this is better as a private var and a getter.